### PR TITLE
Add `BCSS` dataset

### LINKED ISSUE #553: Add `BCSS` segmentation dataset
[The BCSS dataset](https://github.com/CancerDataScience/CrowdsourcingDataset-Amgadetal2019) contains over 20,000 segmentation annotations of tissue region from breast cancer images from TCGA - https://bcsegmentation.grand-challenge.org/

Repo: https://github.com/PathologyDataScience/BCSS?tab=readme-ov-file

### PR DESCRIPTION
**Pull request by nkaenzig**
Closes #553 

After 4 epochs using ImageNet pre-trained `vit_small_patch16_224_dino` encoder: 
image

### CODE REVIEW

**Review by ioangatop — COMMENTED** [review]
looks great, cool that we mix some minor staff as well - the only think that I have to ask of course if for the split, thoughts?

File: src/eva/vision/data/datasets/segmentation/bcss.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+                test_indices.append(i)
+            else:
+                train_val_indices.append(i)
+
+        train_sub_indices, val_sub_indices, _ = splitting.random_split(
+            samples=train_val_indices,
+            train_ratio=self._train_split_ratio,
+            val_ratio=self._val_split_ratio,
+            seed=self._seed,
+        )
```
[the remark this answers is not in the document]
Comment (nkaenzig) [reply src/eva/vision/data/datasets/segmentation/bcss.py:171]: Good point. in this dataset, each .png corresponds to a different TCGA slide, so at slide-level there is no data leakage. Also, all images seem to come from different patients, so also no patient-level data leakage.
However, we have multiple images from the same medical centers. The test set contains only images from centers that are not present in the train set. For the random train/val split we do here, the centers might be overlapping, so there is "leakage" at the center-level, but I'd still consider this a realistic scenario for train/val.

File: src/eva/vision/data/datasets/segmentation/bcss.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    """Validation split ratio."""
+
+    _expected_length: int = 151
+    """Expected dataset length."""
+
+    def __init__(
+        self,
+        root: str,
+        sampler: samplers.Sampler,
+        split: Literal["train", "val", "test"] | None = None,
```
[the remark this answers is not in the document]
Comment (nkaenzig) [reply src/eva/vision/data/datasets/segmentation/bcss.py]: done

File: docs/datasets/bcss.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# BCSS
+
+The BCSS (Breast Cancer Semantic Segmentation) consists of extracts from 151 WSI images from [TCGA](https://www.cancer.gov/ccg/research/genome-sequencing/tcga), containing over 20.000 segmentation annotations covering 21 different tissue types.
```
Comment (roman807) [inline docs/datasets/bcss.md:3]: nit: 20.000 -> 20,000

File: src/eva/vision/data/datasets/segmentation/bcss.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nkaenzig) [reply src/eva/vision/data/datasets/segmentation/bcss.py:173]: Haven't run a lot of experiments yet, I wanted to leave that for when we have a clearer training protocol for segmentation. So far I only validated that it learns the task.

File: docs/datasets/bcss.md (hunk 3 above)
Comment (nkaenzig) [reply docs/datasets/bcss.md]: fixed

File: src/eva/vision/data/datasets/segmentation/consep.py [hunk 4]
```diff
@@ -145,23 +142,11 @@ def _load_file_paths(self, split: Literal["train", "val"] | None = None) -> List
...
         filename = self.filename(index).split(".")[0]
-        mask_dir = "Train" if filename.startswith("train") else "Test"
-        return os.path.join(self._root, mask_dir, "Labels", f"{filename}.mat")
-
-    def _extract_mask_patch(self, index: int, mask: npt.NDArray[Any]) -> npt.NDArray[Any]:
-        """Reads the mask patch at the coordinates corresponding to the given patch index."""
-        (x, y), width, height = self._get_coords(index)
-        return mask[x : x + width, y : y + height]
+        mask_dir = "Train/Labels" if filename.startswith("train") else "Test/Labels"
+        return os.path.join(self._root, mask_dir, f"{filename}.mat")
```
Comment (ioangatop) [inline src/eva/vision/data/datasets/segmentation/consep.py:149]: I think we shouldn't change that,  because windows used for example will have issues, as we have the `/` pre-determined here for the split

so this is safare 
```
mask_dir = "Train" if filename.startswith("train") else "Test"
return os.path.join(self._root, mask_dir, "Labels", f"{filename}.mat")
```

File: src/eva/vision/data/datasets/segmentation/bcss.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ioangatop) [reply src/eva/vision/data/datasets/segmentation/bcss.py:173]: make sense, we can configure it afterwards

File: src/eva/vision/data/datasets/segmentation/bcss.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    def __init__(
+        self,
+        root: str,
+        sampler: samplers.Sampler,
+        split: Literal["train", "val", "trainval", "test"] | None = None,
+        width: int = 224,
+        height: int = 224,
+        target_mpp: float = 0.5,
+        transforms: Callable | None = None,
+        seed: int = 42,
```
Comment (ioangatop) [inline src/eva/vision/data/datasets/segmentation/bcss.py:58]: I would prefere to maybe remove the seed from here and fix the dataset on the 42, and we can control the overall seed of eva from only the global one, what do you think?

File: src/eva/vision/data/datasets/segmentation/bcss.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    _val_split_ratio: float = 0.2
+    """Validation split ratio."""
+
+    _expected_length: int = 151
+    """Expected dataset length."""
+
+    def __init__(
+        self,
+        root: str,
+        sampler: samplers.Sampler,
```
Comment (ioangatop) [inline src/eva/vision/data/datasets/segmentation/bcss.py:52]: it comes handy to have maybe a default value here (like the grid sampler), especially when using the dataset from python. Because its a class, we can do `sampler: samplers.Sampler | None` and the init to set it to the default value if `None`. Any thoughts?

File: src/eva/vision/data/datasets/segmentation/bcss.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+        split: Literal["train", "val", "trainval", "test"] | None = None,
+        width: int = 224,
+        height: int = 224,
```
Comment (ioangatop) [inline src/eva/vision/data/datasets/segmentation/bcss.py:55]: maybe at a different PR but I would expect these values to be at the sampler's `__init__` values, as I believe this is major part of it (on the class side I think it makes a bit more sense to add it in the init args instead of the `sample` method) - thoughts?

File: src/eva/vision/data/datasets/segmentation/bcss.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        target_mpp: float = 0.5,
```
Comment (ioangatop) [inline src/eva/vision/data/datasets/segmentation/bcss.py:56]: this `target_mpp` doesn't do much here right? make we can remove it and move it to the init hardcoded to this value

**Review by ioangatop — COMMENTED** [review]
Looks great! here couple of comments, I think we will merge it super soon

File: src/eva/vision/data/datasets/segmentation/bcss.py (hunk 5 above)
Comment (nkaenzig) [reply src/eva/vision/data/datasets/segmentation/bcss.py:58]: I'll remove the seed & random split here and replace by a static `val` dataset based on a selection of centers as discussed.

File: src/eva/vision/data/datasets/segmentation/bcss.py (hunk 6 above)
Comment (nkaenzig) [reply src/eva/vision/data/datasets/segmentation/bcss.py:52]: good idea, done

File: src/eva/vision/data/datasets/segmentation/bcss.py (hunk 8 above)
Comment (nkaenzig) [reply src/eva/vision/data/datasets/segmentation/bcss.py:56]: I'd like to keep this here. In fact, having this as a flexible option helped me find during experimentation that mpp=0.25 yielded patches that were too small. Maybe users for certain FM evaluations might want to operate at different mpps, so I think it'd be nice to expose this as configurable option.

File: src/eva/vision/data/datasets/segmentation/bcss.py (hunk 7 above)
Comment (nkaenzig) [reply src/eva/vision/data/datasets/segmentation/bcss.py:55]: Good point. On the class side of things I agree it'd be cleaner to have it in the `sampler` init. But from a user perspective, having it exposed at the dataset init might be easier to understand / more intuitive.

No strong preference though, we can implement this in a follow up PR.

File: src/eva/vision/data/datasets/segmentation/bcss.py (hunk 5 above)
Comment (nkaenzig) [reply src/eva/vision/data/datasets/segmentation/bcss.py]: done

File: src/eva/vision/data/datasets/segmentation/bcss.py (hunk 7 above)
Comment (nkaenzig) [reply src/eva/vision/data/datasets/segmentation/bcss.py:60]: Opened an issue for this here: https://github.com/kaiko-ai/eva/issues/568

File: src/eva/vision/data/datasets/segmentation/bcss.py (hunk 8 above)
Comment (nkaenzig) [reply src/eva/vision/data/datasets/segmentation/bcss.py:61]: Probably also makes sense to refactor this together with `width` & `height` in a follow up PR: https://github.com/kaiko-ai/eva/pull/559#discussion_r1678863002

**Review by ioangatop — APPROVED** [review]
